Move view index files to .recovery when db is deleted

jiangphcn opened the pull request:
<!-- Thank you for your contribution!

     Please file this form by replacing the Markdown comments
     with your text. If a section needs no action - remove it.

     Also remember, that CouchDB uses the Review-Then-Commit (RTC) model
     of code collaboration. Positive feedback is represented +1 from committers
     and negative is a -1. The -1 also means veto, and needs to be addressed
     to proceed. Once there are no objections, the PR can be merged by a
     CouchDB committer.

     See: http://couchdb.apache.org/bylaws.html#decisions for more info. -->

## Overview
Before change, the view index files were renamed in place if the corresponding database was deleted and "enable_database_recovery" configuration item is set to true. This allows view index files to be re-used if database is recovered. 
However, these deleted view files are spread widely and  may become orphans. This makes disk management difficult. 
In order to help better manage disk, the change is to implement automatic movement of index files into centralized place when main database is deleted.

<!-- Please give a short brief for the pull request,
     what problem it solves or how it makes things better. -->

## Testing recommendations
make check apps=couch tests=nuke_dir_test_
```
======================== EUnit ========================
Nuke directory tests
  enable_database_recovery = false, context = delete
    couch_file_tests:499: make_rename_dir_test_case...ok
    couch_file_tests:500: make_rename_dir_test_case...ok
    couch_file_tests:501: make_rename_dir_test_case...ok
    [done in 0.009 s]
  enable_database_recovery = false, context = compaction
    couch_file_tests:499: make_rename_dir_test_case...ok
    couch_file_tests:500: make_rename_dir_test_case...ok
    couch_file_tests:501: make_rename_dir_test_case...ok
    [done in 0.009 s]
  enable_database_recovery = true, context = delete
    couch_file_tests:499: make_rename_dir_test_case...ok
    couch_file_tests:500: make_rename_dir_test_case...ok
    couch_file_tests:501: make_rename_dir_test_case...ok
    [done in 0.009 s]
  enable_database_recovery = true, context = compaction
    couch_file_tests:499: make_rename_dir_test_case...ok
    couch_file_tests:500: make_rename_dir_test_case...ok
    couch_file_tests:501: make_rename_dir_test_case...ok
    [done in 0.009 s]
  delete_after_rename = true
    couch_file_tests:517: make_delete_dir_test_case...ok
    couch_file_tests:518: make_delete_dir_test_case...ok
    couch_file_tests:519: make_delete_dir_test_case...ok
    couch_file_tests:520: make_delete_dir_test_case...ok
    [done in 0.012 s]
  delete_after_rename = false
    couch_file_tests:517: make_delete_dir_test_case...ok
    couch_file_tests:518: make_delete_dir_test_case...ok
    couch_file_tests:519: make_delete_dir_test_case...ok
    couch_file_tests:520: make_delete_dir_test_case...ok
    [done in 0.012 s]
  [done in 0.473 s]
=======================================================
  All 20 tests passed.
```
<!-- Describe how we can test your changes.
     Does it provides any behaviour that the end users
     could notice? -->

## GitHub issue number
This PR will be the number
<!-- If this is a significant change, please file a separate issue at:
     https://github.com/apache/couchdb/issues
     and include the number here and in commit message(s) using
     syntax like "Fixes #472" or "Fixes apache/couchdb#472".  -->

## Related Pull Requests

<!-- If your changes affects multiple components in different
     repositories please put links to those pull requests here.  -->

## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [ ] Documentation reflects the changes;

jiangphcn commented:
@eiri Thanks again for your comments. I have addressed them except for item3 "run some kind of perf test". For item2 "the final path for the recovery indexes", you can see some test result below.
`make check apps=couch tests=deleted_filename_test_`. Would you please take a further look?

```
======================== EUnit ========================
couch_file:758: should_create_proper_deleted_filename (/srv/data/dbname.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.dbname_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/shards/00000000-1fffffff/dbname.1458336317.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/dbname.1458336317_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/dbname.1458336317_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/dbname.1499329402_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/dbname.1499329402_design/mrview/8fabddcb28f501d6764afd7def3bd352.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/db.name.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.db.name_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/shards/00000000-1fffffff/db.name.1458336317.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/db.name.1458336317_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/db.name.1458336317_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/db.name.1499329402_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/db.name.1499329402_design/mrview/8fabddcb28f501d6764afd7def3bd352.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/user/dbname.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.user/dbname_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/shards/00000000-1fffffff/user/dbname.1458336317.couch)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/user/dbname.1458336317_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.shards/00000000-1fffffff/user/dbname.1458336317_design/mrview/3133e28517e89a3e11435dd5ac4ad85a.view)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/user/dbname.1499329402_design)...ok
couch_file:758: should_create_proper_deleted_filename (/srv/data/.recovery/shards/00000000-1fffffff/user/dbname.1499329402_design/mrview/8fabddcb28f501d6764afd7def3bd352.view)...ok
=======================================================
  All 21 tests passed.
```

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -264,6 +265,19 @@ rename_file(Original) ->
+rename_dir(RootDelDir, Original, DbName) ->
+    DbDir = binary_to_list(DbName) ++ "_design",
+    [DbPureName | _R] = lists:reverse(filename:split(binary_to_list(DbName))),
+    Deleted_Index_Dir = filename:join(
```
Remove the underscores in variable names.

jiangphcn replied:
changed

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -264,6 +265,19 @@ rename_file(Original) ->
 
+rename_dir(RootDelDir, Original, DbName) ->
+    DbDir = binary_to_list(DbName) ++ "_design",
+    [DbPureName | _R] = lists:reverse(filename:split(binary_to_list(DbName))),
```
What is this logic doing? Its not making sense to me why we'd be looking at just the last element of the DbName.

jiangphcn replied:
The logic of above line is to get pure dbname with epochs suffix. 

Let me give example:
```
RootDelDir = /srv/view_index
DbDir = shards/e0000000-ffffffff/testdb1.1500350972_design
Original = /srv/view_index/.shards/e0000000-ffffffff/testdb1.1500350972_design
DBName = shards/e0000000-ffffffff/testdb1.1500350972
```
In this PR, we experienced two evolutions. 

First, we select recovery directory, like `filename:join([RootDelDir, ".recovery", DbDir]),`. Thus, the constructed directory looks like `/srv/view_index/.recovery/shards/e0000000-ffffffff/testdb1.1500350972_design`. This is feasible. 

However, we want to group these moved view directories and files. So you can see that current approach is `filename:join([RootDelDir, ".recovery", DbPureName, DbDir]),`. The DbPureName looks like `testdb1.1500350972` and doesn't contains `shards/e0000000-ffffffff`. Thus, for view files belonging to the same database, they can be easily managed by using such directory structure. There are two points which are considered:
  - for database which are deleted and created using same name, their epochs are different. This will not cause confusion.
  - for database with the same name from different user, in most situation, their epochs are different. Even if their epochs are the same, they can be still managed because they will be put into different subdirectory, like
  `/srv/view_index/.recovery/testdb1.1500350972/shards/e0000000-ffffffff/user1/testdb1.1500350972_design` and
  `/srv/view_index/.recovery/testdb1.1500350972/shards/e0000000-ffffffff/user2/testdb1.1500350972_design`

davisp replied:
Where was the discussion on grouping them like that? I don't see any immediate reason why the first approach of just mirroring the shards/$range/dbname/_design... hierarchy under the .recovery directory.

And the particular case I was referring to is if your dbname has an embedded / in it which is feasible. Because given your example it reads to me like if we had a dbname <<"foo/bar">> then we'd end up with files under .recovery/bar/shards/... which would be wrong as far as I can tell.

davisp replied:
Also, while users is a Cloudant specific issue it doesn't change the logic here as if you had a database name hierarchy you'd end up just chopping off the last element for the group which seems like it'd be super confusing to any administrator trying to use file recovery. You'd basically have to write the same scripts to search for files as you would without the whole grouping effort which makes me think there's not much benefit to doing it in the first place.

jiangphcn replied:
The discussion was reflected in summary of this commit (https://github.com/apache/couchdb/pull/597/commits/9b1dbfd5a2d770a4f2c2a5cb79c6e4575dad36c1).
```
 - Introduce dbname directory between .recovery and shards so that
   all view files belonging to the same db can be located at the
   the central place. Thus, these view files can be easily sorted
   by atime at /srv/data/.recovery level.
   When these view files from same db needs to be recovered, they
   can be easily moved back to view directory.
```

However, considering the user information in DBName and situation where "/" is included in dbname, I changed back to `/srv/view_index/.recovery/shards/e0000000-ffffffff/testdb1.1500350972_design` approach.

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -264,6 +265,19 @@ rename_file(Original) ->
         Else -> Else
     end.
 
+rename_dir(RootDelDir, Original, DbName) ->
```
This function does not seem to be handling databases that may have multiple path elements (ie, a DbName with a '/' in it) like you do for cloudant-labs/hastings#3.

jiangphcn replied:
The situation where DbName can contains "/" was considered. For details, please see response to next comment.

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -272,14 +286,18 @@ deleted_filename(Original) ->
+        true when Context == delete ->
+            DbName = couch_util:get_value(db_name, Options),
+            rename_dir(RootDelDir, Dir, DbName);
+        true -> rename_file(Dir);
```
Wouldn't this be when Context == compact, thus we'd want to delete the file (via moving to the .delete directory if enabled).

jiangphcn replied:
```
   case EnableRecovery of
        true when Context == delete ->
            DbName = couch_util:get_value(db_name, Options),
            rename_dir(RootDelDir, Dir, DbName);
        true -> rename_file(Dir);
        false -> delete_dir(RootDelDir, Dir)
```
From above, you can see that there are three conditions:
 - the first one: `EnableRecovery == true` and `Context == delete`
    under this situation, we want to move index to .recovery directory
 - the third one: `EnableRecovery == false`
    under this situation, we want to delete the file (either delete in place or move it to .delete directory

If we are talking about the second situation where `EnableRecovery == true` and `Context == compact`,  we still want to keep original logic, i.e. rename view file in place.

davisp replied:
Ah, gotchya. For some reason I was thinking that was somewhere else but I forgot what module i was reading.

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -317,6 +335,9 @@ init_delete_dir(RootDir) ->
     ok.
 
+init_recovery_dir(RootDir) ->
+    Dir = filename:join(RootDir,".recovery"),
```
Missing space after comma here and the next line.

jiangphcn replied:
changed

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -750,11 +771,28 @@ make_filename_fixtures(DbNames) ->
+            % by splitting Format String using ~s.
+            TildeSOccNum = erlang:length(binary:split(
+                list_to_binary(Format),
+                list_to_binary("~s"),
```
Why not just write <<"~s">> here?

jiangphcn replied:
changed

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -750,11 +771,28 @@ make_filename_fixtures(DbNames) ->
         lists:map(fun(Format) ->
-            filename:join("/srv/data", io_lib:format(Format, [DbName]))
+            % calculate the occurrence of "~s" in the Format string
+            % by splitting Format String using ~s.
```
This comment would be better as something like:

> Count how many times we need to specify the database name by splitting on the ~s formatter.

jiangphcn replied:
changed

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -750,11 +771,28 @@ make_filename_fixtures(DbNames) ->
+                list_to_binary("~s"),
+                [global])
+            ) - 1,
+            if TildeSOccNum == 1 ->
```
TildeSOccNum is a fairly confusing name. Something like "ArgCount".

Using `if` probably isn't the cleanest approach here. I'd do something like:

```erlang
Args = case ArgCount of
    1 -> [DbName];
    2 -> [DbName, DbName]
end,
filename:join("/srv/data/", io_lib:format(Format, Args)
```

jiangphcn commented:
@davisp Hi Paul, thanks for your review and comments. I addressed most of comments. For other comments, I leave some response. Would you please check? Thanks again

jiangphcn commented:
@davisp Hi Paul, I removed dbname directory between .recovery and shards in new commit. Would you please take another look? Thanks

davisp commented on src/couch/src/couch_file.erl:
```diff
@@ -264,6 +265,18 @@ rename_file(Original) ->
 
+rename_dir(RootDelDir, Original, DbName) ->
+    DbDir = binary_to_list(DbName) ++ "_design",
+    DeletedIndexDir = filename:join(
```
Minor nit, this should be RenamedIndexDir or future us might get confused why its called Deleted but isn't actually getting deleted.

jiangphcn commented:
Thanks @davisp for your review. I have changed name and could you please take yet another look when getting time? Thanks again.

davisp commented:
+1

wohali commented:
**Please do not merge this to master until after the 2.1 branch is created. This should happen in the next 7 days.**

wohali commented:
You may merge this one when ready. Thanks for waiting! :)

cdlwjing commented:
According to the discussion with @jiangphcn, I tried to run some test cases to compare the performance difference between two designs(the current design and the new design in this PR). From the test result, I can see there are some perf improvement with new design. There is a report with all test details under: ```https://github.com/cloudant/couchdb/blob/index-delete-new-design-testreport/test_report/index_delete_comparision_report.ipynb```

wohali commented:
Go ahead and merge this when ready @jiangphcn .

jiangphcn commented:
Thanks @wohali for your reminder and patience. There is some suggestion from Ops expert about not moving view index to /srv/view_index/.recovery directory. Otherwise, it will use more inode and there is risky that there is no atime to set for some mounted driver. I have to hold on merge this PR and will come back later. Thanks again.

wohali commented:
@jiangphcn it's been a month. If we're still not sure what to do here, could we close this PR and open a new PR in the future instead, when you're ready? Thanks.

jiangphcn commented:
yes, let me close this PR @wohali

jiangphcn commented:
@wohali It looks that I don't have permission to close this PR. Can you help close this PR? Thanks

jiangphcn commented:
thanks @davisp for closing this PR.
